core/field.js

```diff
@@ -492,42 +499,172 @@ Blockly.Field.prototype.updateWidth = function() {
   this.size_.width = width;
 };
 
+/**
+ * Default font returned from CanvasRendering2DContext.font.
+ */
+Blockly.Field.CANVAS_CONTEXT_SPEC_DEFAULT_FONT = '10px sans-serif';
+
+/**
+ * Default tool for measuring the width of some text. Used if a
+ * CanvasRendering2DContext cannot be configured to measure text for the font
+ * of a dom class.
+ */
+Blockly.Field.defaultTextTool = {
+  measureText: function(textElement) {
+    var width;
+
+    // Attempt to compute fetch the width of the SVG text element.
+    try {
+      if (goog.userAgent.IE || goog.userAgent.EDGE) {
+        width = textElement.getBBox().width;
+      } else {
+        width = textElement.getComputedTextLength();
+      }
+    } catch (e) {
+      // In other cases where we fail to geth the computed text.
+      // Instead, use an approximation and do not cache the result. At
+      // some later point in time when the block is inserted into the
+      // visible DOM, this method will be called again and, at that
+      // point in time, will not throw an exception.
+      width = textElement.textContent.length * 8;
+    }
+
+    return {
+      width: width
+    };
+  }
+};
+
+/**
+ * Initialize some objects for measuring text widths.
+ * @param {Array} classNames names of classes to initialize
+ * @param {Element} element relative element to get font values from
+ * @private
+ */
+Blockly.Field.initTextTools = function(classNames, element) {
+  if (!Blockly.Field.cacheTextTools_) {
+    return;
+  }
+
+  if (classNames.length === 1 && (
+    classNames[0] === 'blocklyFlyoutLabelText' ||
+    classNames[0] === 'blocklyText' ||
+    classNames[0] === 'blocklyText blocklyDropdownText'
+  )) {
+    // These three are commonly used so lets initialize them together. This will
+    // reduce the cost of DOM style and layout while determining the font.
+    classNames = [
+      'blocklyFlyoutLabelText',
+      'blocklyText',
+      'blocklyText blocklyDropdownText'
+    ];
+  }
+
+  var root = document.body;
```

Comment:
Can you explain what this root is - I'm confused :)
  I thought getRootNode would return the document, though when I look at this in the debugger, it is returning the <g> tag. Then later you're then adding another <g> and appending to that.  Is that right?

Reply:
This version needs to determine the blockly fonts from CSS. (We could hard code them in JS.) So we need nodes in the DOM to fetch that style. Our common case is to get the fonts for all the blockly text node class names. This root is where we attach those nodes so we can get the font info for each class in one pass.

`document.body` is from a version of this I wrote where the text element may not have been attached to the dom yet. We can change this and require it be an element in the DOM. Which, the element is, in this version.

Adding a node, getting the info, removing it, adding the second, getting the info, removing, etc. will incur a restyle event and performance cost. Reusing the same node would also incur this cost. In the best case, we add all the nodes to the DOM, get the font info for each node, then remove all the nodes from the DOM.

The parent is here to conveniently let us detach the nodes after we get the font info for all of the nodes. We could add and remove the nodes to and from the root without the parent alternatively. As long as we add them all, get all their info, and remove them all.